lib/gceimgutils/gceutils.py

```diff
@@ -161,12 +161,16 @@ def get_version():
 # ----------------------------------------------------------------------------
 def get_compute_api(credentials):
     """Build the compute API"""
+
     global compute_api
 
-    if not compute_api:
-        compute_api = discovery.build('compute', 'v1', credentials=credentials)
+    api = compute_api.get(credentials)
+
+    if not api:
+        api = discovery.build('compute', 'v1', credentials=credentials)
+        compute_api[credentials] = api
```

Comment:
Aren't credentials dictionaries which aren't hashable? Also, is there any advantage to cache this? It seems like use case that includes different projects would create different instance of the utility?

Reply:
This would be a much bigger change but it would be nice for the api to be a lazy attribute of the class. Then the code just uses it and if it's not initialized it does so before use.

Reply:
credentials is an object as such can be used as a key
```
>>> from google.oauth2 import service_account
>>> credentials = service_account.Credentials.from_service_account_file('....json')
>>> a = {credentials: 'foo'}
>>>
```

If I remember correctly Python just uses the address of an object as the key in this case. But it's been a long time since I dug into those details.

The utility is not a class and as such, because of the `global` declaration and the way Python treats name lookup in any given Python process that loads, for example gceremoveimg, even if GCERemoveImage is instantiated multiple times there exists only one copy of gceutils with a global setting of `compute_api` in the gceutils namespace. This is why our automatic deletions have been failing when multiple projects are involved.

Because the build of the api takes a rather long time (> 1.3 s) based on a average of 10 api object creations the construction was moved out into the utility rather than being part of the base class and the caching was implemented. Otherwise every object construction would get the > 1.3 s performance hit for the construction of the api object at some point during it's life-time.

So moving the construction of the API into the base class is an option but we'd take the > 1.3s hit for every object, list img or rm at some point. While in the case of our automation we take the performance hit only once per project we operate on with this implementation. If the number for constructing the API object were not so big having it in the classes that do the operations would be preferable, I agree.

Reply:
Ack as an object the hash should work. However, the hash I assume will always be different because it's going to be a new instance of credentials object if it's a new instance of GCERemoveImage.

IMHO global hashing seems to be solving a downstream project problem and opening up chances to unknowingly use the wrong credentials.